### ISSUE TITLE
Running Graphene Shielded Containers in Marblerun

### ISSUE THREAD

**Issue by ratnadeepb** [edgelesssys/marblerun#187]
### Use case
At present Marblerun allows to run marbles directly. However, I could not any references to running secure containers like the ones created by the Graphene Shielded Container tool:
```
sudo ./gsc build --insecure-args nginx test/ubuntu18.04-python3.manifest
sudo ./gsc sign-image nginx enclave-key.perm
```

### Describe your solution

I would like to know if there is any way or a plan to allow secure containers to run under marblerun.

**Comment by thomasten** [comment]
Hi @ratnadeepb,
we plan to integrate with gsc and have this in our backlog. I can't give a definite ETA yet. Maybe sometime next month.

**Comment by ratnadeepb** [comment]
In the meantime, is there another way to run services in containers on marblerun?

**Comment by m1ghtym0** [comment]
@ratnadeepb You can build your own docker containers with EGo, Graphene, or Occlum services.
That's what gsc is abstracting for you with Graphene.

**Comment by Nirusu** [comment]
You can take a look onto our graphene-redis sample on how to achieve this:
https://github.com/edgelesssys/marblerun/tree/master/samples/graphene-redis

It basically just takes the Redis sample from the Graphene repo, adds the Marblerun premain process to it and creates a docker image from it. That's how you run services in containers with Marblerun :) 

This can actually even be less steps when using the prebuilt premain process as a download instead of building it like in the example Dockerfile.

**Comment by ratnadeepb** [comment]
Yes, I was looking at that example. So if I understand that example correctly, then basically download or build the executable and premain in the EdglessRT container and deploy?

**Comment by ratnadeepb** [comment]
The `redis` example uses this line in the Dockerfile: `RUN --mount=type=secret,id=signingkey,dst=/graphene/Pal/src/host/Linux-SGX/signer/enclave-key.pem,required=true \
    make clean && make SGX=1`
Instead of building a C application in situ, what if a go (not `ego`) application is copied over to the docker. Will it still work?

More importantly, how do I generate the [manifest](https://github.com/edgelesssys/marblerun/blob/master/samples/graphene-redis/redis-manifest.json)? Especially, I don't see a way to run `oesign eradump -e enclave.signed` (from [here](https://www.marblerun.sh/docs/workflows/add-service/#step-21-define-the-enclave-software%20package) in this case?

**Comment by m1ghtym0** [comment]
> The `redis` example uses this line in the Dockerfile: `RUN --mount=type=secret,id=signingkey,dst=/graphene/Pal/src/host/Linux-SGX/signer/enclave-key.pem,required=true \ make clean && make SGX=1`
> Instead of building a C application in situ, what if a go (not `ego`) application is copied over to the docker. Will it still work?
> 

Graphene is working on adding Go support: https://github.com/oscarlab/graphene/issues/396
I'm not sure if it is officially supported yet.

> More importantly, how do I generate the [manifest](https://github.com/edgelesssys/marblerun/blob/master/samples/graphene-redis/redis-manifest.json)? Especially, I don't see a way to run `oesign eradump -e enclave.signed` (from [here](https://www.marblerun.sh/docs/workflows/add-service/#step-21-define-the-enclave-software%20package)) in this case since I am building the application traditionally?

We don't have the tooling to extract enclave-related information from a docker container directly yet.
You'll need to extract the enclave binary from the image (`docker cp`) and obtain the values from there.
Our documentation is missing an explanation of how to do this for Graphene and Occlum workloads. (I've opened a ticket for that)
Graphene stores this information in the sigstruct `.sig` file they generate.
They provide a parser of this file for GSC using `docker cp`: https://github.com/oscarlab/graphene/blob/master/Tools/gsc/gsc.py#L322
But I'm not aware of a tool for regular Graphene enclaves.
We mostly use the triple of `MRSIGNER`+`ProductID`+`SecurityVersion` which are all independent of the enclave binary hence, this wasn't an issue so far. 
Let me check with the Graphene folks what their usual procedure is and I'll get back to you.

**Comment by m1ghtym0** [comment]
@ratnadeepb You can use the [graphene-sgx-get-token](https://github.com/oscarlab/graphene/blob/master/python/graphene-sgx-get-token) to get those values from the .sig file.

**Comment by dimakuv** [comment]
Hi @ratnadeepb and @m1ghtym0 from a Graphene maintainer :)

Accidentally stumbled upon this issue, so I want to leave a couple notes here:
- Google Go is technically supported by Graphene, but we currently don't have an example for it. We will add an example for Google Go apps at some point (in 2-3 months, maybe sooner).
- Graphene will soon have proper installation and packaging. The downloadable packages will be available for Debian/Ubuntu and hopefully for CentOS/Fedora. With this, installing Graphene on the bare system or in a Docker container will be a much more straight-forward task.
- GSC will be soon updated to use Graphene packages instead of building Graphene from sources.
- Graphene has the `graphene-sgx-sign` Python tool to generate the final `manifest.sgx` file and the `.sig` file (contains SIGSTRUCT). GSC inside of it contains another tool `finalize-manifest` that automatically generates `.manifest` file, and then GSC gives this generated `.manifest` file to the `graphene-sgx-sign` tool of Graphene.
